<a href="https://colab.research.google.com/github/MaanasSM/Deep-Learning-Mini-Projects/blob/main/Artificial_Neural_Networks_lab3_1_22_25.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ARTIFICIAL NEURAL NETWORK LAB3


In [ ]:
# Import necessary libraries
from keras.models import Sequential
from keras.layers import Dense, Activation
from keras.utils import to_categorical
import numpy as np
from sklearn import datasets  #it loads a set of datasets
from sklearn.model_selection import train_test_split


# Load the Iris dataset   # this dataset is there in datasets from sklearn, which is already imported
iris = datasets.load_iris()
X, y = datasets.load_iris(return_X_y=True)


# One-hot encode the labels (since it's a multi-class classification problem)
y = to_categorical(y, 3)   # a kind of encoding applied onto the data values


# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.40, random_state=42)
#test size = 0.40 => 40% of sample is for testing, and 60% is for training => model will be trained using 60%, aandd
# the remaining 40% is unknown to the model, this can be used for testing the model accuracy

#actually, normalsation is optional
# Normalize the input features (to have mean=0 and variance=1)
X_train = (X_train - np.mean(X_train, axis=0)) / np.std(X_train, axis=0) # (x-mean)/sd => this is to find the z-score (z-test in probability and statistics), normalising done to test how much the data deviated from mean and all
X_test = (X_test - np.mean(X_test, axis=0)) / np.std(X_test, axis=0)


# Define the neural network model
model = Sequential()


# Input layer and first hidden layer with 10 neurons
model.add(Dense(10, input_shape=(4,)))
model.add(Activation('relu'))  # Using ReLU for hidden layers


# Second hidden layer with 10 neurons
model.add(Dense(10))
model.add(Activation('relu'))


# Output layer with 3 neurons (one for each class)
model.add(Dense(3))
model.add(Activation('softmax'))  # Softmax for multi-class classification

#and thus, a total of 23 hidden layers (23 is splitted into 3 parts => 10+10+3)
#and this complete model will be executed in a sequential way

# Compile the model
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])


# Print a summary of the model
model.summary()
#summary of model will have 23 layers => i/p, 10 layers in the second, 10 layers in third, 3 layers in third, and then output


# Train the model
# model.fit(X_train, y_train, epochs=100, batch_size=32)
model.fit(X_train, y_train, epochs=100, batch_size=32)


# Evaluate the model on the test set
score = model.evaluate(X_test, y_test)


# Print the test accuracy
print(f'Test Loss: {score[0]}')
print(f'Test Accuracy: {score[1]*100:.2f}%')




Model: "sequential_9"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_27 (Dense)                     │ (None, 10)                  │              50 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ activation_27 (Activation)           │ (None, 10)                  │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_28 (Dense)                     │ (None, 10)                  │             110 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ activation_28 (Activation)           │ (None, 10)                  │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_29 (Dense)                     │ (None, 3)                   │              33 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ activation_29 (Activation)           │ (None, 3)                   │               0 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 193 (772.00 B)

 Trainable params: 193 (772.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 2s 208ms/step - accuracy: 0.2727 - loss: 1.6572
Epoch 2/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3235 - loss: 1.5713 
Epoch 3/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.3486 - loss: 1.5940 
Epoch 4/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3330 - loss: 1.5078 
Epoch 5/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3369 - loss: 1.4840 
Epoch 6/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3408 - loss: 1.4496 
Epoch 7/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3274 - loss: 1.4012 
Epoch 8/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3297 - loss: 1.4100 
Epoch 9/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3007 - loss: 1.3833 
Epoch 10/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3108 - loss: 1.3152 
Epoch 11/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.2694 - loss: 1.3247 
Epoch 12/100
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.2929 - lo

Total params: 193 (772.00 B)
 Trainable params: 193 (772.00 B)
 Non-trainable params: 0 (0.00 B)

 means that all the parameters were learnt by the model


improving model accuracy => multiple ways

increase the number of epochs

or change the learning rate


Things called hyper parameter tuning
